### PR TITLE
Use Base in Charmhub package and its response structures.

### PR DESCRIPTION
**Pull request by hmlanigan**
Make the least intrusive changes from platform to base while allowing for base to be used when communicating with the Charmhub API. This is a first step to replacement within the entire juju code base.

Add RefreshBase to correspond with RefreshPlatform.  All translation from RefreshPlatform to RefreshBase happens within the charmhub package.

Translate from a Base in the Charmhub API response structs inside of the charms and charmhub facades.

## QA steps

Check that the changes in charmhub/transport match the starcraft bases spec.

`juju info` and `juju find` will be tested once the appropriate API changes are ready for those calls. `juju info --series focal ubuntu-juju-qa` will not return correct data.  For now, ensure that the calls do not fail, but some data may be missing.

```console
$ juju bootstrap localhost  test --config logging-config="juju.apiserver.charms.client=TRACE; juju.apiserver.charmhub.client=TRACE; juju.resource.resourceadapters.charmhub.client=TRACE; juju.apiserver.charmrevisionupdater.charmhub=TRACE" --build-agent
$ juju add-model six --config charmhub-url="https://api.staging.snapcraft.io"

# ubuntu-juju-qa should deploy to bionic
$ juju deploy ubuntu-juju-qa
$ juju deploy ubuntu --series xenial

# There should be no errors related to installing the resource this config change initiates.
$ juju deploy postgresql --config aws_region="us-east-1"

# review the logs to see that base is used in charmhub api calls and responses.
$ juju debug-log -m controller
```

### CODE REVIEW

File: charmhub/transport/error_test.go [hunk 1]
```diff
@@ -50,8 +50,8 @@ two`)
 func (ErrorSuite) TestExtras(c *gc.C) {
 	expected := APIError{
 		Extra: APIErrorExtra{
-			DefaultPlatforms: []Platform{
-				{Architecture: "amd64", OS: "ubuntu", Series: "focal "},
+			DefaultBases: []Base{
+				{Architecture: "amd64", Name: "ubuntu", Channel: "focal "},
```
Comment (manadart) [inline charmhub/transport/error_test.go:54]: 20.04

File: cmd/juju/charmhub/download.go [hunk 2]
```diff
@@ -332,7 +332,7 @@ func (c *downloadCommand) suggested(ser string, channel string, releases []trans
 	series := set.NewStrings()
 	for _, rel := range releases {
 		if rel.Channel == channel {
-			series.Add(rel.Platform.Series)
+			series.Add(rel.Base.Channel)
```
Comment (manadart) [inline cmd/juju/charmhub/download.go:335]: I think here we would need to convert the channel that we receive back to a series.

File: charmhub/find_test.go [hunk 3]
```diff
@@ -127,10 +127,10 @@ func (s *FindSuite) TestFindRequestPayload(c *gc.C) {
 			DefaultRelease: transport.FindChannelMap{
 				Channel: transport.Channel{
 					Name: "latest/stable",
-					Platform: transport.Platform{
+					Base: transport.Base{
 						Architecture: "all",
-						OS:           "ubuntu",
-						Series:       "bionic",
+						Name:         "ubuntu",
+						Channel:      "bionic",
```
Comment (manadart) [inline charmhub/find_test.go:133]: 18.04

File: apiserver/facades/client/charmhub/convert.go [hunk 4]
```diff
@@ -142,32 +146,44 @@ func transformInfoChannelMap(channelMap []transport.InfoChannelMap) ([]string, m
...
-			Series:       v.Series,
+			OS:           v.Name,
+			Series:       series,
 		}
 	}
 	return out
 }
 
-func convertCharm(info transport.InfoResponse) (*params.CharmHubCharm, []string) {
+func convertCharm(info transport.InfoResponse) (*params.CharmHubCharm, []string, bool) {
```
Comment (manadart) [inline apiserver/facades/client/charmhub/convert.go:171]: We will need to detect whether we are dealing with V2 metadata here and determine the series from bases in manifest.yaml. Not to what extent this is mandatory for 2.9.0.

File: apiserver/facades/client/charmhub/convert.go [hunk 5]
```diff
@@ -142,32 +146,44 @@ func transformInfoChannelMap(channelMap []transport.InfoChannelMap) ([]string, m
...
 
-func convertPlatforms(in []transport.Platform) []params.Platform {
+// convertBasesToPlatforms converts a base to a platform.  Is mean to be a short
+// term solution due to schedule.  It should be removed once platforms are
+// replaced with bases through out the code.
+func convertBasesToPlatforms(in []transport.Base, isKub bool) []params.Platform {
 	out := make([]params.Platform, len(in))
 	for i, v := range in {
+		channel := v.Channel
+		if isKub {
```
Comment (manadart) [inline apiserver/facades/client/charmhub/convert.go:156]: Is the direction mixed up here? We *do* want to store series "kubernetes" in this case right?
```go
	for i, v := range in {
		var series string
		if isKub {
			series = "kubernetes"
		} else {
			series, _ = coreseries.VersionSeries(v.Channel)
		}
		out[i] = params.Platform{
			Architecture: v.Architecture,
			OS:           v.Name,
			Series:       series,
		}
	}
```

**Review by manadart — COMMENTED** [review]
Some initial comments. I see changes were incoming as I was making them.

File: cmd/juju/charmhub/download.go (hunk 2 above)
Comment (hmlanigan) [reply cmd/juju/charmhub/download.go:335]: Something else is going on, downloadSuite.TestRunWithUnsupportedSeries is checking for series to be listed and passes.  Therefore something else is going on, as your comment should be true based on the line.  Will investigate.

File: cmd/juju/charmhub/download_test.go [hunk 6]
```diff
@@ -263,31 +263,31 @@ func (s *downloadSuite) expectRefreshUnsupportedSeries() {
 				Extra: transport.APIErrorExtra{
 					Releases: []transport.Release{
 						{
-							Platform: transport.Platform{Architecture: "amd64", OS: "ubuntu", Series: "bionic"},
-							Channel:  "stable",
+							Base:    transport.Base{Architecture: "amd64", Name: "ubuntu", Channel: "bionic"},
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply cmd/juju/charmhub/download_test.go:266]: done

File: charmhub/transport/error_test.go (hunk 1 above)
Comment (hmlanigan) [reply charmhub/transport/error_test.go:54]: done

File: charmhub/find_test.go (hunk 3 above)
Comment (hmlanigan) [reply charmhub/find_test.go:133]: done and others

File: apiserver/facades/client/charmhub/convert.go (hunk 4 above)
Comment (hmlanigan) [reply apiserver/facades/client/charmhub/convert.go:171]: I'll add a TODO to isKubernetes()

File: cmd/juju/charmhub/download.go (hunk 2 above)
Comment (hmlanigan) [reply cmd/juju/charmhub/download.go:335]: done

File: apiserver/facades/client/charmhub/convert.go (hunk 5 above)
Comment (hmlanigan) [reply apiserver/facades/client/charmhub/convert.go:156]: Yes, thank you. And I need to use the series, to get the OS, or we'll send up with something like Ubuntu Kubernetes.

File: charmhub/refresh.go [hunk 7]
```diff
@@ -417,25 +438,40 @@ func (c refreshMany) String() string {
 	return strings.Join(plans, "\n")
 }
 
-// constructRefreshPlatform creates a refresh request platform that allows for
-// partial platform queries.
-func constructRefreshPlatform(platform RefreshPlatform) (transport.Platform, error) {
+// constructRefreshBase creates a refresh request base that allows for
+// partial base queries.
+func constructRefreshBase(platform RefreshPlatform) (transport.Base, error) {
```
Comment (manadart) [inline charmhub/refresh.go:443]: Nit: `makeRefreshBase` for convention.

**Review by manadart — APPROVED** [review]
I am approving for now because we are OK to land this on the 2.9-bases branch, but there is one issue.

If I bootstrap to Microk8s instead, I can't deploy anything because they have channel=kubernetes instead of a base for the latest LTS as we agreed.

Looks like we also need to update the headers that we're sending.
```
controller-0: 17:28:22 TRACE juju.apiserver.charms.client POST request POST /v2/charms/refresh HTTP/1.1
Host: api.staging.snapcraft.io
Accept: application/json
Content-Type: application/json
User-Agent: Juju/2.9-rc9
X-Juju-Metadata: series=kubernetes
X-Juju-Metadata: arch=amd64
X-Juju-Metadata: os=kubernetes

{"context":[],"actions":[{"action":"install","instance-key":"8d8ec827-036b-429b-86f8-abd33a57ae0f","name":"jupyter-ui","channel":"stable","base":{"architecture":"amd64","name":"kubernetes","channel":"20.04"}}]}

controller-0: 17:28:22 TRACE juju.apiserver.charms.client POST response HTTP/1.1 200 OK
Content-Length: 1272
Content-Type: application/json
Date: Fri, 09 Apr 2021 15:28:22 GMT
Server: gunicorn/19.7.1
Snap-Store-Version: 35
X-Request-Id: 57F4FF33EC5B0A324F5701BB607072961AD0
X-Vcs-Revision: 46660a0
X-View-Name: snapdevicegw.webapi_refresh_charm.charm_refresh

{"error-list":[],"results":[{"charm":{"id":"uOlZjTga1livQhKwrvyQ3YNYOr09Ym0Y","license":"","name":"jupyter-ui","publisher":{"display-name":"Notebook Charmers","id":"2arNltAu0kDncLEoTBv7rkJp3KNXqJTL","username":"notebook-charmers","validation":"unproven"},"summary":"Multi-user server for Jupyter notebooks","type":"charm"},"error":{"code":"revision-not-found","extra":{"releases":[{"base":{"architecture":"amd64","channel":"kubernetes","name":"kubernetes"},"channel":"stable","platform":{"architecture":"amd64","os":"kubernetes","series":"kubernetes"}},{"base":{"architecture":"amd64","channel":"kubernetes","name":"kubernetes"},"channel":"candidate","platform":{"architecture":"amd64","os":"kubernetes","series":"kubernetes"}},{"base":{"architecture":"amd64","channel":"kubernetes","name":"kubernetes"},"channel":"beta","platform":{"architecture":"amd64","os":"kubernetes","series":"kubernetes"}},{"base":{"architecture":"amd64","channel":"kubernetes","name":"kubernetes"},"channel":"edge","platform":{"architecture":"amd64","os":"kubernetes","series":"kubernetes"}}]},"message":"No revision was found in the Store."},"id":"uOlZjTga1livQhKwrvyQ3YNYOr09Ym0Y","instance-key":"8d8ec827-036b-429b-86f8-abd33a57ae0f","name":"jupyter-ui","released-at":null,"result":"error"}]}

controller-0: 17:28:22 TRACE juju.apiserver.charms.client Refresh() unmarshalled: []transport.RefreshResponse{
    {
        Entity: transport.RefreshEntity{
            Type:      "charm",
            Download:  transport.Download{},
            ID:        "uOlZjTga1livQhKwrvyQ3YNYOr09Ym0Y",
            License:   "",
            Name:      "jupyter-ui",
            Publisher: {"display-name":"Notebook Charmers", "id":"2arNltAu0kDncLEoTBv7rkJp3KNXqJTL", "username":"notebook-charmers", "validation":"unproven"},
            Resources: nil,
            Revision:  0,
            Summary:   "Multi-user server for Jupyter notebooks",
            Version:   "",
            CreatedAt: time.Time{},
        },
        EffectiveChannel: "",
        Error:            &transport.APIError{
            Code:    "revision-not-found",
            Message: "No revision was found in the Store.",
            Extra:   transport.APIErrorExtra{
                Releases: {
                    {
                        Base:    transport.Base{Architecture:"amd64", Name:"kubernetes", Channel:"kubernetes"},
                        Channel: "stable",
                    },
                    {
                        Base:    transport.Base{Architecture:"amd64", Name:"kubernetes", Channel:"kubernetes"},
                        Channel: "candidate",
                    },
                    {
                        Base:    transport.Base{Architecture:"amd64", Name:"kubernetes", Channel:"kubernetes"},
                        Channel: "beta",
                    },
                    {
                        Base:    transport.Base{Architecture:"amd64", Name:"kubernetes", Channel:"kubernetes"},
                        Channel: "edge",
                    },
                },
                DefaultBases: nil,
            },
        },
        ID:          "uOlZjTga1livQhKwrvyQ3YNYOr09Ym0Y",
        InstanceKey: "8d8ec827-036b-429b-86f8-abd33a57ae0f",
        Name:        "jupyter-ui",
        Result:      "error",
        ReleasedAt:  time.Time{},
    },
}
```
We'll have to follow up with them.

**Comment by hmlanigan** [comment]
Made a card for follow up https://trello.com/c/AMn7h0I6/3602-resolve-issue-deploying-k8s-charm-with-bases

**Comment by hmlanigan** [comment]
$$merge$$

**Comment by hmlanigan** [comment]
$$merge$$

**Comment by hmlanigan** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
I found kubeflow tests are failing: `ERROR refresh: unexpected api error attempting to query charm or bundle from the charmhub store
`
Not sure if it's related to this change.

```console
$ juju download kubeflow --debug --channel stable
10:21:11 INFO  juju.cmd supercommand.go:56 running juju [2.9-rc9 0 285c60df453db31a5ad4bb66aec63ac22b1f2c33 gc go1.14.15]
10:21:11 DEBUG juju.cmd supercommand.go:57   args: []string{"juju", "download", "kubeflow", "--debug", "--channel", "stable"}
10:21:11 INFO  juju.juju api.go:78 connecting to API addresses: [localhost:38245 10.152.183.202:17070]
10:21:11 DEBUG juju.api apiclient.go:1132 successfully dialed "wss://localhost:44067/model/7d753a09-d11e-4992-88cc-93ea6ec4c39d/api"
10:21:11 INFO  juju.api apiclient.go:1034 cannot resolve "localhost": operation was canceled
10:21:11 INFO  juju.api apiclient.go:664 connection established to "wss://localhost:44067/model/7d753a09-d11e-4992-88cc-93ea6ec4c39d/api"
10:21:11 INFO  juju.juju api.go:328 API endpoints changed from [localhost:38245 10.152.183.202:17070] to [localhost:44067 10.152.183.202:17070]
10:21:11 DEBUG juju.api monitor.go:35 RPC connection died
10:21:11 INFO  juju.juju api.go:78 connecting to API addresses: [localhost:44067 10.152.183.202:17070]
10:21:11 DEBUG juju.api apiclient.go:1132 successfully dialed "wss://localhost:43665/model/7d753a09-d11e-4992-88cc-93ea6ec4c39d/api"
10:21:11 INFO  juju.api apiclient.go:1034 cannot resolve "localhost": operation was canceled
10:21:11 INFO  juju.api apiclient.go:664 connection established to "wss://localhost:43665/model/7d753a09-d11e-4992-88cc-93ea6ec4c39d/api"
10:21:11 INFO  juju.juju api.go:328 API endpoints changed from [localhost:44067 10.152.183.202:17070] to [localhost:43665 10.152.183.202:17070]
10:21:11 DEBUG juju.api monitor.go:35 RPC connection died
10:21:13 INFO  cmd download.go:411 Response HTTP/1.1 400 BAD REQUEST
Content-Length: 126
Content-Type: application/json
Date: Mon, 12 Apr 2021 00:21:13 GMT
Server: gunicorn/19.7.1
Snap-Store-Version: 34
X-Request-Id: 79D102F994140A325C4A01BB60739278C46AC19
X-Vcs-Revision: 61bdb1d
X-View-Name: snapdevicegw.webapi_refresh_charm.charm_refresh

{"error-list":[{"code":"api-error","message":"Additional properties are not allowed ('base' was unexpected) at /actions/0"}]}

10:21:13 INFO  cmd download.go:411 charmhub API error api-error:Additional properties are not allowed ('base' was unexpected) at /actions/0
ERROR unexpected api error attempting to query charm or bundle from the charmhub store
10:21:13 DEBUG cmd supercommand.go:537 error stack:
github.com/juju/juju/charmhub/errors.go:73: unexpected api error attempting to query charm or bundle from the charmhub store
github.com/juju/juju/charmhub/refresh.go:137:
github.com/juju/juju/cmd/juju/charmhub/download.go:254:
```

**Comment by hmlanigan** [comment]
@ycliuhw, we're working in a feature branch because things will break until all of the changes are made.  @manadart has already found one issue we need to resolve above.  Are you using 2.9?  Or is this from the CI runs?

**Pull request merged**